### PR TITLE
Kickstart contract

### PR DESCRIPTION
**Pull request by barroqt**
-Add funding allocation
-Reward distribution from admin
-TODO: handle dates

### CODE REVIEW

File: contracts/Kickstart.sol [hunk 1]
```diff
@@ -0,0 +1,69 @@
+//SPDX-License-Identifier: MIT
+pragma solidity >=0.7.0 <0.9.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "hardhat/console.sol";
+
+contract Kickstart {
```
Comment (toji59) [inline contracts/Kickstart.sol]: Contract should be called `Investment` and be used as part of the factory.

File: contracts/Kickstart.sol [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    IERC20 public token;
+
+    // TODO: allocation percentage: mapping(address => uint)
+
+    constructor(uint256 _fundingGoal) {
+        investor = msg.sender;
+        fundingGoal = _fundingGoal;
+        console.log(investor.balance);
+        balances[investor] = investor.balance;
+        token = IERC20(0xA048B6a5c1be4b81d99C3Fd993c98783adC2eF70); // AVAX address
```
Comment (toji59) [inline contracts/Kickstart.sol]: We should have the token address as an input

File: contracts/Kickstart.sol [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    mapping(address => bool) public hasInvested;
+
+    IERC20 public token;
+
+    // TODO: allocation percentage: mapping(address => uint)
+
+    constructor(uint256 _fundingGoal) {
+        investor = msg.sender;
+        fundingGoal = _fundingGoal;
+        console.log(investor.balance);
```
Comment (toji59) [inline contracts/Kickstart.sol]: `console.log` to remove

File: contracts/Kickstart.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        require(msg.sender == admin, "Only the administrator can do that");
+        _;
+    }
+
+    function depositAllocation(uint _amount, address _to) external payable {
+        require(balances[msg.sender] >= _amount, "Not enough tokens");
+        require(!hasInvested[msg.sender], "This user already invested");
+        require(totalAllocatedAmount + _amount <= fundingGoal, "Can't invest more than the targeted amount");
+        // require(now > startDate, "The campaign has not started yet");
+        // require(now < endDate, "The campaign is over");
```
Comment (ngrt) [inline contracts/Kickstart.sol:53]: we should use the `block.timestamp` enable to time check feature

File: contracts/Kickstart.sol [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+        hasInvested[msg.sender] = true; // We know this address is eligible for rewards
+        investedAmount[msg.sender] = _amount; // We know the amount invested by this address
+        totalAllocatedAmount += _amount; // The amount of token invested for this campaign
+        balances[msg.sender] -= _amount;
```
Comment (toji59) [inline contracts/Kickstart.sol]: Shouldn't we get the token from the ERC20 token contract? I would use the `transferFrom()` from IERC20 `token`. Let me know your thoughts.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol#L63

Comment (barroqt) [reply contracts/Kickstart.sol]: You're right we should do that

File: contracts/Investment.sol [hunk 6]
```diff
@@ -1,18 +1,72 @@
 //SPDX-License-Identifier: MIT
-pragma solidity ^0.8.0;
+pragma solidity >=0.7.0 <0.9.0;
 
 import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "hardhat/console.sol";
```
Comment (toji59) [inline contracts/Investment.sol]: Remove this import

File: contracts/Investment.sol [hunk 7]
```diff
@@ -1,18 +1,72 @@
...
 
-    uint256 public x;
+    // User status
+    address public admin;
+    address public investor;
 
-    function init(uint256 _x) external initializer {
-        x = _x;
+    // keeps track of processed campaigns
+    uint256 campaignId; // Do we need this?
```
Comment (toji59) [inline contracts/Investment.sol]: What would be the usage? Can we use it for the `round`?

Comment (toji59) [reply contracts/Investment.sol]: Let's have an offline discussion

File: contracts/Investment.sol [hunk 8]
```diff
@@ -1,18 +1,72 @@
...
+    // An Allocation has an Investor 
+    mapping(address => uint256) public investedAmount;
+    // who already invested to avoid investing twice
+    mapping(address => bool) public hasInvested;
+
+    IERC20 public token;
+
+    // TODO: allocation percentage: mapping(address => uint)
+   
+   // use timestamp startDate:1648196138 (2022-03-25 09:15:38) endDate:1648624538 (2022-03-30 09:15:38) for passing later checks
```
Comment (toji59) [inline contracts/Investment.sol]: Comment to remove

File: contracts/Investment.sol [hunk 9]
```diff
@@ -1,18 +1,72 @@
...
+    modifier onlyAdmin() {
+        require(msg.sender == admin, "Only the administrator can do that");
+        _;
+    }
+
+    function depositAllocation(uint _amount, address _to) external payable {
+        require(balances[msg.sender] >= _amount, "Not enough tokens");
+        require(!hasInvested[msg.sender], "This user already invested");
+        require(totalAllocatedAmount + _amount <= fundingGoal, "Can't invest more than the targeted amount");
+        require(block.timestamp > startDate, "The campaign has not started yet");
```
Comment (toji59) [inline contracts/Investment.sol]: `block.timestamp >= startDate`

File: contracts/Investment.sol [hunk 10]
```diff
@@ -1,18 +1,72 @@
...
+        require(block.timestamp < endDate, "The campaign is over");
+
+        hasInvested[msg.sender] = true; // We know this address is eligible for rewards
+        investedAmount[msg.sender] = _amount; // We know the amount invested by this address
+        totalAllocatedAmount += _amount; // The amount of token invested for this campaign
+        token.transferFrom(msg.sender, _to, _amount);
+    }
+    
+    function sendRewards(uint _amount, address _to) external onlyAdmin() {
```
Comment (toji59) [inline contracts/Investment.sol]: Here we might need to have a `sendRewards` on the Investment level. Goal would be to allow the admin to distribute to all the allocation owner using one call. There are tradeoff to it as we will probably have to loop through the allocations.

Comment (toji59) [reply contracts/Investment.sol]: also I would call the method `distributeToken`.

Comment (barroqt) [reply contracts/Investment.sol]: But when the investor deposits tokens, the campaign is still ongoing, we won't know the reward allocation before the end date as anyone could participate and therefore impact the reward %

**Pull request merged**